### PR TITLE
Add documentation for  template.openshift.io/expose annotation

### PR DESCRIPTION
**Pull request by jim-minter**
Will be arriving in enterprise 3.6.

### CODE REVIEW

**Comment by jim-minter** [comment]
@bparees @adellape fyi

**Comment by bparees** [comment]
lgtm

**Comment by bparees** [comment]
@openshift/team-documentation ptal/merge

File: dev_guide/templates.adoc [hunk 1]
```diff
@@ -347,34 +347,53 @@ labels:
...
 
 === Description
-The template description covers information that informs users what your
-template does and helps them find it when searching in the web console. In
-addition to general descriptive information, it includes a set of tags. Useful
-tags include the name of the language your template is related to (e.g., *java*,
-*php*, *ruby*, etc.).
+The template description informs users what the template does and helps them
+find it when searching in the web console. Additional metadata beyond the
+template name is optional, but is typically useful to have. In addition to
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: s/but is typically useful to have/but useful

File: dev_guide/templates.adoc [hunk 2]
```diff
@@ -347,34 +347,53 @@ labels:
...
+general descriptive information, the metadata also includes a set of tags.
+Useful tags include the name of the language the template is related to (e.g.,
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: s/e.g.,/for example,
s/etc./and so on.

File: dev_guide/templates.adoc [hunk 3]
```diff
@@ -396,11 +415,15 @@ customizations] that can be added to an {product-title} cluster that uses your
 template. You must specify an icon class that exists, or it will prevent falling
 back to the generic icon.
 endif::[]
-<6> An instructional message that is displayed when this template is instantiated.
-This field should inform the user how to use the newly created resources.
-Parameter substitution is performed on the message before being displayed so
-that generated credentials and other parameters can be included in the output.
-Include links to any next-steps documentation that users should follow.
+<7> The name of the person or organisation providing the template.
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: organization

File: dev_guide/templates.adoc [hunk 4]
```diff
@@ -609,6 +632,85 @@ namespace the parameter substitution resolved the value to, assuming the user
 has permission to create objects in that namespace.
 ====
 
+[[writing-exposing-object-fields]]
+
+=== Exposing Object Fields
+
+Template authors may indicate that fields of particular objects in a template
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: s/may/can
s/recognises/recognizes

File: dev_guide/templates.adoc [hunk 5]
```diff
@@ -609,6 +632,85 @@ namespace the parameter substitution resolved the value to, assuming the user
...
+should be exposed. The template service broker recognises exposed fields on
+ConfigMap, Secret, Service and Route objects and returns the values of the
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: objects, and

File: dev_guide/templates.adoc [hunk 6]
```diff
@@ -609,6 +632,85 @@ namespace the parameter substitution resolved the value to, assuming the user
...
+exposed fields when a user binds a service backed by the broker.
+
+To expose one or more fields of an object, add annotations prefixed by
+`template.openshift.io/expose-` or `template.openshift.io/base64-expose-` to
+the object in the template. The annotation key following the prefix is passed
+through to key the returned field in the bind response.
+
+Use the `template.openshift.io/expose-` annotation to return the field value as
+a string. This is convenient although it does not handle arbitrary binary data.
+If you wish to return binary data, the `template.openshift.io/base64-expose-`
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: s/wish to/want to

Comment (ahardin-rh) [inline dev_guide/templates.adoc]: use the the `template.openshift.io/base64-expose-` annotation instead to base64 encode the data before it is returned.

File: dev_guide/templates.adoc [hunk 7]
```diff
@@ -609,6 +632,85 @@ namespace the parameter substitution resolved the value to, assuming the user
...
+annotation can be used instead to base64 encode the data before it is returned.
+
+The value of the expose annotation is a [Kubernetes JSONPath expression]
+(https://kubernetes.io/docs/user-guide/jsonpath/) which is resolved at bind time
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: is a link:https://kubernetes.io/docs/user-guide/jsonpath/[Kubernetes JSONPath expression], which is resolved

File: dev_guide/templates.adoc [hunk 8]
```diff
@@ -609,6 +632,85 @@ namespace the parameter substitution resolved the value to, assuming the user
...
+- kind: Route
+  apiVersion: v1
+  metadata:
+    annotations:
+      template.openshift.io/expose-uri: "http://{.spec.host}{.spec.path}"
+  spec:
+    path: mypath
+====
+
+An example response to a Bind operation given the above partial template
```
Comment (ahardin-rh) [inline dev_guide/templates.adoc]: `bind` operation

**Comment by jim-minter** [comment]
@ahardin-rh all changes pushed.  Is there a house style guide anywhere for future reference?

**Comment by ahardin-rh** [comment]
@jim-minter Thanks! Yes, we have these [doc guidelines ](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc) and also go by the IBM Style Guide. :beers:

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../dev_guide/templates.adoc#dev-guide-templates[Templates]
|Added a new xref:../dev_guide/templates.adoc#writing-exposing-object-fields[Exposing Object Fields] section.
%

**Pull request merged**